### PR TITLE
store/tikv: tiny update mocktikv

### PR DESCRIPTION
**Pull request by tiancaiamao**
1. for mocktikv, close storage should  close the leveldb engine
2. set tikvStore's mock flag by NewMockTikvStore function

Point 1 will benefit this PR https://github.com/pingcap/tidb/pull/4543 , to be more precisely: fix.

For point 2, check `mock` by type assert to a specific struct is unreliable:

```
_, mock := client.(*mocktikv.RPCClient)
```

The client maybe anything, such as:

```
type T struct{} // T implements Client interface.
func hijack(c Client) Client {
    return &T{}
}
NewMockTikvStore(WithHijackClient(hjiack))
```

@hanfei1991 @coocood

### CODE REVIEW

File: store/tikv/kv.go [hunk 1]
```diff
@@ -276,6 +282,12 @@ func (s *tikvStore) Close() error {
 	if err := s.client.Close(); err != nil {
 		return errors.Trace(err)
 	}
+
+	// For mocktikv, Close() should not only close the client,
+	// but also the underlying store such as leveldb.
+	if raw, ok := s.client.(realClose); ok {
```
Comment (shenli) [inline store/tikv/kv.go]: Check tikvStore.mock == true?

File: store/tikv/kv.go [hunk 2]
```diff
@@ -262,6 +262,12 @@ func (s *tikvStore) GetSnapshot(ver kv.Version) (kv.Snapshot, error) {
 	return snapshot, nil
 }
 
+type realClose interface {
```
Comment (shenli) [inline store/tikv/kv.go]: Add comment for this interface.

File: store/tikv/kv.go [hunk 3]
```diff
@@ -276,6 +282,12 @@ func (s *tikvStore) Close() error {
 	if err := s.client.Close(); err != nil {
 		return errors.Trace(err)
 	}
+
+	// For mocktikv, Close() should not only close the client,
+	// but also the underlying store such as leveldb.
+	if raw, ok := s.client.(realClose); ok {
+		raw.RealClose()
```
Comment (shenli) [inline store/tikv/kv.go]: check the err

**Comment by coocood** [comment]
@tiancaiamao 
Can we just simply close leveldb in mocktikv Client?

**Comment by tiancaiamao** [comment]
Close leveldb in mocktikv client is not confirm to the close semantics: maybe there are multiple clients and other client is still working. 
The `RealClose` method is necessary. @coocood

File: store/tikv/kv.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/kv.go]: Check realClose interface is more precise than mock flag.
If  `tikvStore.mock == true`, maybe it's not leveldb engine and doesn't need to close.
If `client.(realClose)` , RealClose must be called.

**Comment by coocood** [comment]
@tiancaiamao 
In the case of mocktikv, there is only one client for a store.
The `RealClose` method is confusing and doesn't have much semantic difference than `Close`.

File: store/tikv/mock-tikv/rpc.go [hunk 4]
```diff
@@ -605,5 +606,12 @@ func (c *RPCClient) SendReq(ctx goctx.Context, addr string, req *tikvrpc.Request
 
 // Close closes the client.
 func (c *RPCClient) Close() error {
+	if raw, ok := c.MvccStore.(closeable); ok {
```
Comment (coocood) [inline store/tikv/mock-tikv/rpc.go]: Why not just use `io.Closer`?

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**